### PR TITLE
Feature: vscode impl nightlies download and installation

### PR DESCRIPTION
**Pull request by Veetaha**
I need to test things more, but the core shape of the code is quite well-formed.
The main problem is that we save the release date only for nightlies and there are no means to get the release date of the stable extension (i.e. for this we would need to consult the github releases via a network request, or we would need to somehow save this info into package.json or any other file accessible from the extension code during the deployment step, but this will be very hard I guess).
So there is an invariant that the users can install nightly only from our extension and they can't do it manually, because when installing the nightly `.vsix` we actually save its release date to `globalState`

Closes: #3402

TODO:
- [x] More manual tests and documentation

cc @matklad @lnicola

### CODE REVIEW

File: editors/code/src/installation/interfaces.ts [hunk 1]
```diff
@@ -42,6 +41,9 @@ export namespace ArtifactSource {
          */
         repo: GithubRepo;
 
+
+        // FIXME: add installationPath: string;
```
Comment (Veetaha) [inline editors/code/src/installation/interfaces.ts]: This will simplify code a bit, will do it in separate PR in order not to bloat this one.

File: editors/code/src/installation/extension.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    return userResponse === "Download now";
+}
+
+/**
+ * Shutdowns the process in case of success (i.e. reloads the window) or throws an error.
+ *
+ * ACHTUNG!: this function has a crazy amount of state transitions, handling errors during
+ * each of them would result in a ton of code (especially accounting for cross-process
+ * shared mutable `globalState` access). Enforcing reentrancy for this is best-effort.
+ */
```
Comment (Veetaha) [inline editors/code/src/installation/extension.ts]: The ideal code for this function blows my mind. It will require cancellation support along with file-system locking, error state recovery and whatever...

File: editors/code/src/main.ts [hunk 3]
```diff
@@ -34,7 +35,13 @@ export async function activate(context: vscode.ExtensionContext) {
 
     const config = new Config(context);
 
-    const serverPath = await ensureServerBinary(config.serverSource);
+    vscode.workspace.onDidChangeConfiguration(() => ensureProperExtensionVersion(config).catch(log.error));
+
+    // Don't await the user response here, otherwise we will block the lsp server bootstrap
+    void ensureProperExtensionVersion(config).catch(log.error);
+
+    const serverPath = await ensureServerBinary(config);
```
Comment (Veetaha) [inline editors/code/src/main.ts]: I am not entirely sure this is the right decision, since the user may end up looking at both notifications simultaneously:
image

File: editors/code/src/config.ts [hunk 4]
```diff
@@ -128,23 +148,47 @@ export class Config {
...
                 owner: "rust-analyzer",
             }
         };
     }
 
+    get nightlyVsixSource(): ArtifactSource.GithubRelease {
+        return this.createGithubReleaseSource("rust-analyzer.vsix", NIGHTLY_TAG);
+    }
+
+    readonly installedNightlyExtensionReleaseDate = new DateStorage(
```
Comment (matklad) [inline editors/code/src/config.ts]: We can get the date of nightly extension (.vsix, not the server binary) from version in package.json. We set version to date:

https://github.com/rust-analyzer/rust-analyzer/blob/34bc0f4f55a9c2239d77b550b9bbd61b4eef6bb1/xtask/src/dist.rs#L30-L36

File: editors/code/src/main.ts (hunk 3 above)
Comment (matklad) [reply editors/code/src/main.ts]: I think that's [good enough](https://b1.culture.ru/c/793937.jpg)

File: editors/code/src/config.ts (hunk 4 above)
Comment (Veetaha) [reply editors/code/src/config.ts]: Yeah, but that date contains only the day, but no hours, minutes...
This means we will need to keep the invariant of issuing releases not more often than 1 time per day and some more trickery with 25 hours heuristics I guess

File: editors/code/src/installation/extension.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+const HEURISTIC_NIGHTLY_RELEASE_PERIOD_IN_HOURS = 25;
+
+/**
+ * Installs `stable` or latest `nightly` version or does nothing if the current
+ * extension version is what's needed according to `desiredUpdateChannel`.
+ */
+export async function ensureProperExtensionVersion(config: Config): Promise<never | void> {
+    // User has built lsp server from sources, she should manage updates manually
+    if (config.serverPath !== null) return;
```
Comment (Veetaha) [inline editors/code/src/installation/extension.ts]: Just a note: I think this is a bit confusing to coexist with `serverSource`, the better check would be
```suggestion
    if (config.serverSource?.type === ArtifactSource.Type.ExplicitPath) return;
```

File: editors/code/src/installation/server.ts [hunk 6]
```diff
@@ -55,38 +50,53 @@ export async function ensureServerBinary(source: null | ArtifactSource): Promise
...
+        return required.tag !== installed.tag;
+    }
+
+    assert(required.date !== null, "Extension release date should have been saved during its installation");
+    assert(installed.date !== null, "Server release date should have been saved during its installation");
+
+    return installed.date.getTime() !== required.date.getTime();
+}
+
+async function downloadServer(
```
Comment (Veetaha) [inline editors/code/src/installation/server.ts]: Note: this should be wrapped into `utils.notReentrant()` (i.e. this will prevent the user invoking concurrent downloads from rust-analyzer.showServerVersion command)

**Comment by Veetaha** [comment]
@matklad, did some manual testing, and from the naive glance things tend to work fine. However, I couldn't test this in real world scenario, since the nightly version that is currently on our Github releases doesn't contain the code of this PR.

File: editors/code/src/installation/extension.ts [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+/**
+ * Shutdowns the process in case of success (i.e. reloads the window) or throws an error.
+ *
+ * ACHTUNG!: this function has a crazy amount of state transitions, handling errors during
+ * each of them would result in a ton of code (especially accounting for cross-process
+ * shared mutable `globalState` access). Enforcing no reentrancy for this is best-effort.
+ */
+const tryDownloadNightlyExtension = notReentrant(async function tryDownloadNightlyExtension(
```
Comment (matklad) [inline editors/code/src/installation/extension.ts]:
```suggestion
const tryDownloadNightlyExtension = notReentrant(async ()
```

That way, direct recursive calls will also go via wrapper.

Comment (Veetaha) [reply editors/code/src/installation/extension.ts]: I intentionally used a named lambda here, because, we log `fn.name` when reentrancy invariant is broken. Or do you think this won't be necessary as we could follow the stacktrace?

Comment (matklad) [reply editors/code/src/installation/extension.ts]: "decorating" named function in JS is a footgun, as recursive calls (or nested closures) will use undercoated name. This problems seems worse than slight disadvantage in reading the stack trace.

But no strong opinion here. TBH, I would just try very hard to make all control flow here blindly obvious, such that these checks are not required.

**Comment by matklad** [comment]
bors d+

File: editors/code/src/installation/extension.ts (hunk 7 above)
Comment (Veetaha) [reply editors/code/src/installation/extension.ts]: Oh, I just understood what you meant by *"direct recursive calls"* and you do have a point!

Okay, let's not rush and strive to make the control flow more evident.
Do you think it is sufficient to handle reentrancy closer to `ensure*()` functions' call site? 

From my perspective, it would be better to explicitly add cancellation support for `ensureProperExtensionVersion()` and not rely on undocumented vscode behavior as [stated here](https://github.com/rust-analyzer/rust-analyzer/blob/9789f984cbe20a7fca1354e542be7bb345d795eb/editors/code/src/installation/extension.ts#L92-L95).

Besides that I guess there is no need for `downloadServer()` cancellation support right now as we call this function only once, so `notReentrant()` for it is really a debug assertion, whilst in case of `tryDownloadNightlyExtension()` it do can blow at runtime and we handle this by throwing `AssertionError` in my current PR :sweat_smile:, thus my cancellation support proposal.

Comment (matklad) [reply editors/code/src/installation/extension.ts]: I think the current solution is fine. I do believe that the best way to arrive at the good code is to actually rush something working in production, and then start incrementally improving it. I'd rather get this in some form in today's release, the behavior shouldn't be significantly different unless the user explicitly requests nightly. 

The code in question is not in my cache right now, so I can't give specific advice, but I'll look closely at it later today :)

Comment (Veetaha) [reply editors/code/src/installation/extension.ts]: Okay, thanks, I'll change this to arrow functions and merge for now then!

**Comment by Veetaha** [comment]
The current impl is not prefect, but gives some ground for improvements in close future!
bors r+
